crates/wasmtime/src/runtime/gc/enabled/structref.rs

```diff
@@ -0,0 +1,757 @@
+        let me: &Self = unsafe { mem::transmute(index) };
+
+        // Assert we really are just a newtype of a `GcRootIndex`.
+        assert!(matches!(
+            me,
+            Self {
+                inner: GcRootIndex { .. },
+            }
+        ));
+
+        me
+    }
+}
+
+impl StructRef {
+    /// Allocate a new `struct` and get a reference to it.
+    ///
+    /// # Errors
+    ///
+    /// If the allocation cannot be satisfied because the GC heap is currently
+    /// out of memory, but performing a garbage collection might free up space
+    /// such that retrying the allocation afterwards might succeed, then a
+    /// [`GcHeapOutOfMemory<()>`][crate::GcHeapOutOfMemory] error is returned.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the allocator, or any of the field values, is not associated
+    /// with the given store.
+    pub fn new(
+        mut store: impl AsContextMut,
+        allocator: &StructRefPre,
+        fields: &[Val],
+    ) -> Result<Rooted<StructRef>> {
+        Self::_new(store.as_context_mut().0, allocator, fields)
+    }
+
+    pub(crate) fn _new(
+        store: &mut StoreOpaque,
+        allocator: &StructRefPre,
+        fields: &[Val],
+    ) -> Result<Rooted<StructRef>> {
+        assert_eq!(
+            store.id(),
+            allocator.store_id,
+            "attempted to use a `StructRefPre` with the wrong store"
+        );
+
+        // Type check the given values against the field types.
+        let expected_len = allocator.ty.fields().len();
+        let actual_len = fields.len();
+        ensure!(
+            actual_len == expected_len,
+            "expected {expected_len} fields, got {actual_len}"
+        );
+        for (ty, val) in allocator.ty.fields().zip(fields) {
+            assert!(
+                val.comes_from_same_store(store),
+                "field value comes from the wrong store",
+            );
+            let ty = match ty.element_type() {
+                StorageType::I8 | StorageType::I16 => &ValType::I32,
+                StorageType::ValType(ty) => ty,
+            };
+            val.ensure_matches_ty(store, &ty)
+                .context("field type mismatch")?;
+        }
+
+        // Allocate the struct and write each field value into the appropriate
+        // offset.
+        let structref = store
+            .gc_store_mut()?
+            .alloc_uninit_struct(allocator.type_index(), &allocator.layout())
+            .err2anyhow()
+            .context("unrecoverable error when allocating new `structref`")?
+            .ok_or_else(|| GcHeapOutOfMemory::new(()))
+            .err2anyhow()?;
+        let mut store = AutoAssertNoGc::new(store);
+        for (index, (ty, val)) in allocator.ty.fields().zip(fields).enumerate() {
+            structref.initialize_field(
+                &mut store,
+                allocator.layout(),
+                ty.element_type(),
+                index,
+                val.clone(),
+            )?;
```

Comment:
I haven't gotten far enough in the diff so this may be more apparent later, but the `?` here is a bit suspect while there's uninitialized memory above. What's the consequence of returning an error here with uninitialized memory? Is that ok insofar as the actual reference won't go anywhere and it'll get cleaned up on the next GC?

Reply:
> Is that ok insofar as the actual reference won't go anywhere and it'll get cleaned up on the next GC?

That would be the case for tracing GCs. For the DRC collector, I think we would actually permanently leak it, since it's refcount would be stuck at `1`.

So, good eye. I'll handle these errors locally and free the allocation before bailing out of the function.